src/mlpack/core/optimizers/cmaes/cmaes_impl.hpp

```diff
@@ -0,0 +1,942 @@
+   * returned value and therefore rather reSampleSingle() should be used.
+   * @param x Solution vector that gets sampled a new value. If x == NULL new
+   *          memory is allocated and must be released by the user using
+   *          delete[].
+   * @return A pointer to the resampled solution vector, equals input x for
+   *         x != NULL on input.
+   */
+  template<typename funcType, typename T>
+  T* CMAES<funcType, T>::sampleSingleInto(T* x)
+  {
+    if (!x)
+      x = new T[N];
+    addMutation(x, 1.0);
+    return x;
+  }
+
+  /**
+   * Can be called after samplePopulation() to resample single solutions. In
+   * general, the function can be used to sample as many independent
+   * mean+sigma*Normal(0,C) distributed vectors as desired.
+   * @param x Element of the return value of samplePopulation(), that is
+   *          pop[0..\f$\lambda\f$]. This solution vector of the population gets
+   *          sampled a new value.
+   * @return A pointer to the resampled "population" member.
+   */
+  template<typename funcType, typename T>
+  T const* CMAES<funcType, T>::reSampleSingleOld(T* x)
+  {
+    assert(x && "reSampleSingleOld(): Missing input x");
+    addMutation(x, 1.0);
+    return x;
+  }
+
+  /**
+   * Used to reevaluate a slightly disturbed solution for an uncertaintly
+   * measurement. In case if x == NULL on input, the memory of the returned x
+   * must be released.
+   * @param x Solution vector that gets sampled a new value. If x == NULL new
+   *          memory is allocated and must be released by the user using
+   *          delete[] x.
+   * @param pxmean Mean vector \f$\mu\f$ for perturbation.
+   * @param eps Scale factor \f$\epsilon\f$ for perturbation:
+   *            \f$x \sim \mu + \epsilon \sigma N(0,C)\f$.
+   * @return A pointer to the perturbed solution vector, equals input x for
+   *         x != NULL.
+   */
+  template<typename funcType, typename T>
+  T* CMAES<funcType, T>::perturbSolutionInto(T* x, T const* pxmean, T eps)
+  {
+    if (!x)
+      x = new T[N];
+    assert(pxmean && "perturbSolutionInto(): pxmean was not given");
+    addMutation(x, eps);
+    return x;
+  }
+
+  /**
+   * Core procedure of the CMA-ES algorithm. Sets a new mean value and estimates
+   * the new covariance matrix and a new step size for the normal search
+   * distribution.
+   * @param fitnessValues An array of \f$\lambda\f$ function values.
+   * @return Mean value of the new distribution.
+   */
+  template<typename funcType, typename T>
+  T* CMAES<funcType, T>::updateDistribution(const T* fitnessValues)
+  {
+    bool diag = diagonalCov == 1 || diagonalCov >= gen;
+
+    assert(state != UPDATED && "updateDistribution(): You need to call "
+          "samplePopulation() before update can take place.");
+    assert(fitnessValues && "updateDistribution(): No fitness function value array input.");
+
+    if (state == SAMPLED) // function values are delivered here
```

Comment:
this is actually according to my previous test case in which the user must call samplePopulation() 
and after that update distribution to calculate the new mean and shift the population and after words testforTermination() to see if criterias are matched or not